manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShortestPathPlanningTest.scala

```diff
@@ -0,0 +1,93 @@
+        |       (oliver:Person {name: 'Oliver Stone'}),
+        |       (rob:Person {name: 'Rob Reiner'}),
+        |
+        |       (wallStreet:Movie {title: 'Wall Street'}),
+        |       (charlie)-[:ACTED_IN {role: "Bud Fox"}]->(wallStreet),
+        |       (martin)-[:ACTED_IN {role: "Carl Fox"}]->(wallStreet),
+        |       (michael)-[:ACTED_IN {role: "Gordon Gekko"}]->(wallStreet),
+        |       (oliver)-[:DIRECTED]->(wallStreet),
+        |
+        |       (thePresident:Movie {title: 'The American President'}),
+        |       (martin)-[:ACTED_IN {role: "A.J. MacInerney"}]->(thePresident),
+        |       (michael)-[:ACTED_IN {role: "President Andrew Shepherd"}]->(thePresident),
+        |       (rob)-[:DIRECTED]->(thePresident)""",
+        "CREATE INDEX ON :Person(name)"
+    )
+    synopsis("Shortest path finding in Cypher and how it is planned")
+    section("-=-=-=-=-=-=-=-") {
+      p( """Planning shortest paths in Cypher can lead to different query plans depending on the predicates that need
+        |to be evaluated. Internally, Neo4j will use a fast bidirectional breadth-first search algorithm if the
+        |predicates can be evaluated whilst searching for the path. If the predicates need to inspect the whole path
+        |before deciding on whether it is valid or not, this fast algorithm cannot be used, and Neo4j will fall back to
+        |a slower exhaustive search for the path. The difference between these two can be in the order of magnitude,
+        |so it is important to ensure that the fast approach is used for time critical queries.""")
+      p( """When the exhaustive search is planned, it is still only executed when the fast algorithm fails to find any
+           |matching paths. The fast algorithm is always executed first, since it is possible that it can find a valid
+           |path even though that could not be guaranteed at planning time.""")
+      query(
+        """MATCH (martin:Person {name:"Martin Sheen"} ),
+          |      (charlie:Person {name:"Charlie Sheen"}),
+          |      p = shortestPath( (martin)-[rels*]-(charlie) )
+          |WHERE ALL(r in rels WHERE type(r) = "ACTED_IN")
+          |RETURN p""", assertShortestPathLength) {
+        p(
+          """This query can be evaluated with the fast algorithm - there are no predicates that need to see the whole path before being evaluated.""")
+        profileExecutionPlan()
+      }
+      section("Single shortest path with predicates") {
+        p( """Predicates used in the WHERE clause that apply to the shortest path pattern are evaluated before deciding what the shortest matching path is. """)
+        query(
+          """MATCH (charlie:Person {name:"Charlie Sheen"}),
+            |      (martin:Person {name:"Martin Sheen"}),
+            |      p = shortestPath( (charlie)-[*]-(martin) )
+            |WHERE length(p) > 1
+            |RETURN p""", assertShortestPathLength) {
+          p(
```

Comment:
There's a bunch of these unnecessary newlines and spaces and triple-quoted strings. They stick in my eyes, but aren't critical obviously.